PSReadLine/Completion.cs

```diff
@@ -483,30 +592,117 @@ private void PossibleCompletionsImpl(CommandCompletion completions, bool menuSel
                     {
                         undo = true;
                         processingKeys = false;
+                        _visualSelectionCommandCount = 0;
+                        _singleton._mark = savedUserMark;
                     }
-                    else
+                    else if (nextKey == Keys.Delete)
                     {
-                        PrependQueuedKeys(nextKey);
+                        // Esc alternative:
+                        // leave string beginning as in current replacement but don't replace to full completion
+                        if (_current == completions.ReplacementIndex)
+                        {
+                            undo = true;
+                        }
+                        else
+                        {
+                            CompletionResult r = new CompletionResult(matches[selectedItem].CompletionText.Substring(0, _current - completions.ReplacementIndex));
+                            DoReplacementForCompletion(r, completions);
+                        }
                         processingKeys = false;
+                        _visualSelectionCommandCount = 0;
+                        _singleton._mark = savedUserMark;
+                        Render();
                     }
-
-                    if (selectedItem != previousItem)
+                    else if (nextKey == Keys.Space || nextKey == Keys.Enter ||
+                        (doneCompletionKeys.ContainsKey(matches[selectedItem].ResultType) &&
```

Comment:
I'm building with C# 7 now, so maybe introduce a local function so you can avoid the ContainsKey call and just use TryGetValue instead.